src/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentDictionary.cs

```diff
@@ -93,31 +93,31 @@ private static bool IsValueWriteAtomic()
             // See http://www.ecma-international.org/publications/files/ECMA-ST/Ecma-335.pdf
             //
             Type valueType = typeof(TValue);
-            if (valueType.IsEnum)
+            if (!valueType.IsValueType)
             {
-                valueType = Enum.GetUnderlyingType(valueType);
+                return true;
             }
-            bool isAtomic =
-                !valueType.GetTypeInfo().IsValueType ||
-                valueType == typeof(bool) ||
-                valueType == typeof(char) ||
-                valueType == typeof(byte) ||
-                valueType == typeof(sbyte) ||
-                valueType == typeof(short) ||
-                valueType == typeof(ushort) ||
-                valueType == typeof(int) ||
-                valueType == typeof(uint) ||
-                valueType == typeof(float);
-
-            if (!isAtomic && IntPtr.Size == 8)
+
+            switch (Type.GetTypeCode(valueType))
             {
-                isAtomic =
-                    valueType == typeof(double) ||
-                    valueType == typeof(long) ||
-                    valueType == typeof(ulong);
+                case TypeCode.Boolean:
+                case TypeCode.Byte:
+                case TypeCode.Char:
+                case TypeCode.Int16:
+                case TypeCode.Int32:
+                case TypeCode.SByte:
+                case TypeCode.Single:
+                case TypeCode.UInt16:
+                case TypeCode.UInt32:
+                    return true;
+                case TypeCode.Int64:
+                case TypeCode.DateTime:
```

Comment:
Sure. I very much doubt it would be that common, and I added it on the basis that once we're switching on type code adding that case is cheap. If there's any doubt around it I'd rather be cautious.